File: typescript/lld/marketplace/food-delivery/OrderService.ts
import { Courier } from './Courier';
import { MatchingService } from './MatchingService';
import { Order } from './Order';

// Facade — coordinates both sub-machines and the matching service. Kicks
// off courier search on PREPARING, not READY — the whole point of the design.
export class OrderService {
    private orders = new Map<string, Order>();
    private couriers = new Map<string, Courier>();
    private matching = new MatchingService();
    private nextOrderId = 1;

    registerCourier(courier: Courier): void {
        courier.status = 'AVAILABLE';
        this.couriers.set(courier.id, courier);
    }

    placeOrder(customerId: string, restaurantId: string, restaurantLat: number, restaurantLng: number): Order {
        const order = new Order(`o${this.nextOrderId++}`, customerId, restaurantId, restaurantLat, restaurantLng);
        this.orders.set(order.id, order);
        return order;
    }

    // Called the moment the kitchen starts cooking — NOT when the food is
    // ready. This is what makes the two timelines overlap.
    onKitchenStartsPreparing(order: Order, respond: (courierId: string) => boolean): void {
        order.startPreparing();
        const candidates = [...this.couriers.values()]
            .filter(c => c.status === 'AVAILABLE')
            .sort((a, b) => this.distance(a, order) - this.distance(b, order));
        this.matching.findAndClaim(order, candidates, respond);
    }

    onKitchenReady(order: Order): void { order.markReady(); }
    onCourierArrived(order: Order): void { order.courierArrived(); }
    completePickup(order: Order): void { order.pickUp(); }
    trackingStatus(order: Order): string { return order.trackingStatus(); }

    private distance(courier: Courier, order: Order): number {
        return Math.hypot(courier.lat - order.restaurantLat, courier.lng - order.restaurantLng);
    }
}

File: python/lld/marketplace/food-delivery/order_service.py
import math
from courier import CourierAvailability
from matching_service import MatchingService
from order import Order


class OrderService:
    """Facade — coordinates both sub-machines and the matching service.
    Kicks off courier search on PREPARING, not READY — the whole point of
    the design."""

    def __init__(self):
        self._orders = {}
        self._couriers = {}
        self._matching = MatchingService()
        self._next_order_id = 1

    def register_courier(self, courier) -> None:
        courier.status = CourierAvailability.AVAILABLE
        self._couriers[courier.id] = courier

    def place_order(self, customer_id: str, restaurant_id: str, restaurant_lat: float, restaurant_lng: float) -> Order:
        order = Order(f"o{self._next_order_id}", customer_id, restaurant_id, restaurant_lat, restaurant_lng)
        self._next_order_id += 1
        self._orders[order.id] = order
        return order

    def on_kitchen_starts_preparing(self, order, respond) -> None:
        """Called the moment the kitchen starts cooking — NOT when the
        food is ready. This is what makes the two timelines overlap."""
        order.start_preparing()
        candidates = [c for c in self._couriers.values() if c.status == CourierAvailability.AVAILABLE]
        candidates.sort(key=lambda c: self._distance(c, order))
        self._matching.find_and_claim(order, candidates, respond)

    def on_kitchen_ready(self, order) -> None:
        order.mark_ready()

    def on_courier_arrived(self, order) -> None:
        order.courier_arrived()

    def complete_pickup(self, order) -> None:
        order.pick_up()

    def tracking_status(self, order) -> str:
        return order.tracking_status()

    def _distance(self, courier, order) -> float:
        return math.hypot(courier.lat - order.restaurant_lat, courier.lng - order.restaurant_lng)